nextgen: CDC none maintainer node can't replicate data downstream

### What did you do?

1. Deploy a cluster with 3 cdc node, a downstream tidb cluster, a `keyspace1` tidb.
2. Run the following sql on `keyspace1` tidb
  - `create database workload`
  - `use workload`
  - `create table fruit(id bigint not null auto_increment primary key, name varchar(20)not null default '');`
  - `insert into fruit(name) values('apple');`

### What did you expect to see?

The database and table data can be replicate downstream. 

### What did you see instead?

Got an error:
```
2025/09/23 15:21:22.491 +08:00] [ERROR] [event_broker.go:949] ["register table to schemaStore failed"] [dispatcherID=39109221610400561678151131026755554213] [tableID=9] [startTs=461009604568678407] [span="tableID: 9, startKey: 7800000174800000ff00000000095f7200fe, endKey: 7800000174800000ff00000000095f7300fe, keyspaceID: 1\n"] [error="[CDC:ErrInvalidKeyspace]bad keyspace, please match the pattern \"^[a-zA-Z0-9]+(\\-[a-zA-Z0-9]+)*$\", the length should no more than %d, eg, \"simple-keyspace-test\""]
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
(paste TiCDC version here)
```

--- a/cmd/cdc/cli/cli_changefeed_create.go
+++ b/cmd/cdc/cli/cli_changefeed_create.go
@@ -26,7 +26,6 @@ import (
 	"github.com/pingcap/ticdc/cmd/cdc/factory"
 	"github.com/pingcap/ticdc/cmd/util"
 	apiv2client "github.com/pingcap/ticdc/pkg/api/v2"
-	"github.com/pingcap/ticdc/pkg/common"
 	"github.com/pingcap/ticdc/pkg/config"
 	"github.com/pingcap/ticdc/pkg/filter"
 	putil "github.com/pingcap/ticdc/pkg/util"
@@ -122,7 +121,7 @@ func newCreateChangefeedOptions(commonChangefeedOptions *changefeedCommonOptions
 // flags related to template printing to it.
 func (o *createChangefeedOptions) addFlags(cmd *cobra.Command) {
 	o.commonChangefeedOptions.addFlags(cmd)
-	cmd.PersistentFlags().StringVarP(&o.keyspace, "keyspace", "k", common.DefaultKeyspace, "Replication task (changefeed) Keyspace")
+	cmd.PersistentFlags().StringVarP(&o.keyspace, "keyspace", "k", "", "Replication task (changefeed) Keyspace")
 	cmd.PersistentFlags().StringVarP(&o.changefeedID, "changefeed-id", "c", "", "Replication task (changefeed) ID")
 	cmd.PersistentFlags().BoolVarP(&o.disableGCSafePointCheck, "disable-gc-check", "", false, "Disable GC safe point check")
 	cmd.PersistentFlags().Uint64Var(&o.startTs, "start-ts", 0, "Start ts of changefeed")

--- a/cmd/cdc/cli/cli_changefeed_pause.go
+++ b/cmd/cdc/cli/cli_changefeed_pause.go
@@ -19,7 +19,6 @@ import (
 	"github.com/pingcap/ticdc/cmd/cdc/factory"
 	"github.com/pingcap/ticdc/cmd/util"
 	apiv2client "github.com/pingcap/ticdc/pkg/api/v2"
-	"github.com/pingcap/ticdc/pkg/common"
 	"github.com/spf13/cobra"
 )
 
@@ -39,7 +38,7 @@ func newPauseChangefeedOptions() *pauseChangefeedOptions {
 // addFlags receives a *cobra.Command reference and binds
 // flags related to template printing to it.
 func (o *pauseChangefeedOptions) addFlags(cmd *cobra.Command) {
-	cmd.PersistentFlags().StringVarP(&o.keyspace, "keyspace", "k", common.DefaultKeyspace, "Replication task (changefeed) Keyspace")
+	cmd.PersistentFlags().StringVarP(&o.keyspace, "keyspace", "k", "", "Replication task (changefeed) Keyspace")
 	cmd.PersistentFlags().StringVarP(&o.changefeedID, "changefeed-id", "c", "", "Replication task (changefeed) ID")
 	_ = cmd.MarkPersistentFlagRequired("changefeed-id")
 }

--- a/logservice/schemastore/schema_store.go
+++ b/logservice/schemastore/schema_store.go
@@ -261,7 +261,7 @@ func (s *schemaStore) getKeyspaceSchemaStore(keyspaceID uint32) (*keyspaceSchema
 		return nil, errors.Trace(err)
 	}
 
-	if err := s.RegisterKeyspace(context.Background(), keyspaceMeta.Name); err != nil {
+	if err := s.RegisterKeyspace(ctx, keyspaceMeta.Name); err != nil {
 		return nil, errors.Trace(err)
 	}
 

--- a/logservice/schemastore/schema_store.go
+++ b/logservice/schemastore/schema_store.go
@@ -245,13 +245,38 @@ func (s *schemaStore) Name() string {
 
 func (s *schemaStore) getKeyspaceSchemaStore(keyspaceID uint32) (*keyspaceSchemaStore, error) {
 	s.keyspaceLocker.RLock()
-	defer s.keyspaceLocker.RUnlock()
 	store, ok := s.keyspaceSchemaStoreMap[keyspaceID]
 	if ok {
+		s.keyspaceLocker.RUnlock()
 		return store, nil
 	}
+	s.keyspaceLocker.RUnlock()
 
-	return nil, errors.ErrInvalidKeyspace
+	ctx := context.Background()
+
+	// if the schemastore does not contain the keyspace, it means the current
+	// cdc node is a follower node. It should register the keyspace when it try
+	// to get keyspace schema_store
+	keyspaceManager := appcontext.GetService[keyspace.KeyspaceManager](appcontext.KeyspaceManager)
+	keyspaceMeta, err := keyspaceManager.GetKeyspaceByID(ctx, keyspaceID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	err = s.RegisterKeyspace(context.Background(), keyspaceMeta.Name)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	s.keyspaceLocker.RLock()
+	store, ok = s.keyspaceSchemaStoreMap[keyspaceID]
+	if ok {
+		s.keyspaceLocker.RUnlock()
+		return store, nil
+	}
+	s.keyspaceLocker.RUnlock()
+
+	return nil, errors.ErrKeyspaceNotFound
 }
 
 func (s *schemaStore) initialize(ctx context.Context) {

--- a/logservice/schemastore/schema_store.go
+++ b/logservice/schemastore/schema_store.go
@@ -254,9 +254,8 @@ func (s *schemaStore) getKeyspaceSchemaStore(keyspaceID uint32) (*keyspaceSchema
 
 	ctx := context.Background()
 
-	// if the schemastore does not contain the keyspace, it means the current
-	// cdc node is a follower node. It should register the keyspace when it try
-	// to get keyspace schema_store
+	// If the schemastore does not contain the keyspace, it means it is not a maintainer node.
+	// It should register the keyspace when it try to get keyspace schema_store.
 	keyspaceManager := appcontext.GetService[keyspace.KeyspaceManager](appcontext.KeyspaceManager)
 	keyspaceMeta, err := keyspaceManager.GetKeyspaceByID(ctx, keyspaceID)
 	if err != nil {

--- a/logservice/schemastore/schema_store.go
+++ b/logservice/schemastore/schema_store.go
@@ -246,11 +246,10 @@ func (s *schemaStore) Name() string {
 func (s *schemaStore) getKeyspaceSchemaStore(keyspaceID uint32) (*keyspaceSchemaStore, error) {
 	s.keyspaceLocker.RLock()
 	store, ok := s.keyspaceSchemaStoreMap[keyspaceID]
+	s.keyspaceLocker.RUnlock()
 	if ok {
-		s.keyspaceLocker.RUnlock()
 		return store, nil
 	}
-	s.keyspaceLocker.RUnlock()
 
 	ctx := context.Background()
 
@@ -262,18 +261,16 @@ func (s *schemaStore) getKeyspaceSchemaStore(keyspaceID uint32) (*keyspaceSchema
 		return nil, errors.Trace(err)
 	}
 
-	err = s.RegisterKeyspace(context.Background(), keyspaceMeta.Name)
-	if err != nil {
+	if err := s.RegisterKeyspace(context.Background(), keyspaceMeta.Name); err != nil {
 		return nil, errors.Trace(err)
 	}
 
 	s.keyspaceLocker.RLock()
 	store, ok = s.keyspaceSchemaStoreMap[keyspaceID]
+	s.keyspaceLocker.RUnlock()
 	if ok {
-		s.keyspaceLocker.RUnlock()
 		return store, nil
 	}
-	s.keyspaceLocker.RUnlock()
 
 	return nil, errors.ErrKeyspaceNotFound
 }

--- a/logservice/schemastore/schema_store.go
+++ b/logservice/schemastore/schema_store.go
@@ -247,10 +247,11 @@ func (s *schemaStore) getKeyspaceSchemaStore(keyspaceID uint32) (*keyspaceSchema
 	s.keyspaceLocker.RLock()
 	defer s.keyspaceLocker.RUnlock()
 	store, ok := s.keyspaceSchemaStoreMap[keyspaceID]
-	if !ok {
-		return nil, errors.ErrInvalidKeyspace
+	if ok {
+		return store, nil
 	}
-	return store, nil
+
+	return nil, errors.ErrInvalidKeyspace
 }
 
 func (s *schemaStore) initialize(ctx context.Context) {

--- a/logservice/txnutil/lock_resolver.go
+++ b/logservice/txnutil/lock_resolver.go
@@ -71,14 +71,14 @@ func (r *resolver) Resolve(ctx context.Context, keyspaceID uint32, regionID uint
 	})
 
 	keyspaceManager := appcontext.GetService[keyspace.KeyspaceManager](appcontext.KeyspaceManager)
-	keyspaceMeta := keyspaceManager.GetKeyspaceByID(keyspaceID)
-	if keyspaceMeta == nil {
-		return cerror.ErrInvalidKeyspace
+	keyspaceMeta, err := keyspaceManager.GetKeyspaceByID(ctx, keyspaceID)
+	if err != nil {
+		return cerror.Trace(err)
 	}
 
 	storage, err := keyspaceManager.GetStorage(keyspaceMeta.Name)
 	if err != nil {
-		return err
+		return cerror.Trace(err)
 	}
 	kvStorage := storage.(tikv.Storage)
 

--- a/maintainer/testutil/test_util.go
+++ b/maintainer/testutil/test_util.go
@@ -157,3 +157,7 @@ func (m *MockPDAPIClient) Healthy(ctx context.Context, endpoint string) error {
 func (m *MockPDAPIClient) LoadKeyspace(ctx context.Context, keyspace string) (*keyspacepb.KeyspaceMeta, error) {
 	return nil, nil
 }
+
+func (m *MockPDAPIClient) GetKeyspaceMetaByID(ctx context.Context, keyspaceID uint32) (*keyspacepb.KeyspaceMeta, error) {
+	return nil, nil
+}

--- a/pkg/common/span_op.go
+++ b/pkg/common/span_op.go
@@ -20,6 +20,8 @@ import (
 	"github.com/pingcap/kvproto/pkg/keyspacepb"
 	"github.com/pingcap/log"
 	"github.com/pingcap/ticdc/heartbeatpb"
+	"github.com/pingcap/ticdc/pkg/config/kerneltype"
+	cerror "github.com/pingcap/ticdc/pkg/errors"
 	"github.com/pingcap/tidb/pkg/kv"
 	"github.com/pingcap/tidb/pkg/meta/metadef"
 	"github.com/pingcap/tidb/pkg/tablecodec"
@@ -109,10 +111,13 @@ func getTableRange(tableID int64) (startKey, endKey []byte) {
 func GetKeyspaceTableRange(keyspaceID uint32, tableID int64) (startKey, endKey []byte, err error) {
 	startKey, endKey = getTableRange(tableID)
 
-	// If the the keyspaceID is 0, that means we are in the classic mode
-	// fallback to the original table range
-	if keyspaceID == DefaultKeyspaceID {
+	if kerneltype.IsClassic() {
 		return startKey, endKey, nil
+	}
+
+	// DefaultKeyspaceID is not a valid keyspaceID on Next Gen
+	if keyspaceID == DefaultKeyspaceID {
+		return startKey, endKey, errors.Trace(cerror.ErrKeyspaceIDInvalid)
 	}
 
 	// The tikv.NewCodecV2 method requires a keyspace meta

--- a/pkg/errors/error.go
+++ b/pkg/errors/error.go
@@ -361,8 +361,12 @@ var (
 		errors.RFCCodeText("CDC:ErrInvalidKeyspace"),
 	)
 	ErrKeyspaceNotFound = errors.Normalize(
-		`keyspace not found`,
+		"keyspace not found",
 		errors.RFCCodeText("CDC:ErrKeyspaceNotFound"),
+	)
+	ErrKeyspaceIDInvalid = errors.Normalize(
+		"keyspace id is invalid",
+		errors.RFCCodeText("CDC:ErrKeyspaceIDInvalid"),
 	)
 	ErrInvalidEtcdKey = errors.Normalize(
 		"invalid key: %s",

--- a/pkg/errors/error.go
+++ b/pkg/errors/error.go
@@ -360,6 +360,10 @@ var (
 		`bad keyspace, please match the pattern "^[a-zA-Z0-9]+(\-[a-zA-Z0-9]+)*$", the length should no more than %d, eg, "simple-keyspace-test"`,
 		errors.RFCCodeText("CDC:ErrInvalidKeyspace"),
 	)
+	ErrKeyspaceNotFound = errors.Normalize(
+		`keyspace not found`,
+		errors.RFCCodeText("CDC:ErrKeyspaceNotFound"),
+	)
 	ErrInvalidEtcdKey = errors.Normalize(
 		"invalid key: %s",
 		errors.RFCCodeText("CDC:ErrInvalidEtcdKey"),

--- a/pkg/eventservice/event_broker.go
+++ b/pkg/eventservice/event_broker.go
@@ -925,6 +925,7 @@ func (c *eventBroker) addDispatcher(info DispatcherInfo) error {
 	err := c.schemaStore.RegisterTable(span.KeyspaceID, span.GetTableID(), info.GetStartTs())
 	if err != nil {
 		log.Error("register table to schemaStore failed",
+			zap.Uint32("keyspaceID", span.KeyspaceID),
 			zap.Stringer("dispatcherID", id), zap.Int64("tableID", span.GetTableID()),
 			zap.Uint64("startTs", info.GetStartTs()), zap.String("span", common.FormatTableSpan(span)),
 			zap.Error(err),

--- a/pkg/keyspace/keyspace_manager.go
+++ b/pkg/keyspace/keyspace_manager.go
@@ -68,8 +68,8 @@ func (k *keyspaceManager) LoadKeyspace(ctx context.Context, keyspace string) (*k
 	}
 
 	k.keyspaceMu.Lock()
-	defer k.keyspaceMu.Unlock()
 	meta := k.keyspaceMap[keyspace]
+	k.keyspaceMu.Unlock()
 	if meta != nil {
 		return meta, nil
 	}
@@ -88,6 +88,13 @@ func (k *keyspaceManager) LoadKeyspace(ctx context.Context, keyspace string) (*k
 		return nil, errors.Trace(err)
 	}
 
+	k.keyspaceMu.Lock()
+	defer k.keyspaceMu.Unlock()
+	// Double check, another goroutine might have fetched and stored it.
+	if meta, ok := k.keyspaceMap[keyspace]; ok {
+		return meta, nil
+	}
+
 	k.keyspaceMap[keyspace] = meta
 	k.keyspaceIDMap[meta.Id] = meta
 
@@ -102,8 +109,8 @@ func (k *keyspaceManager) GetKeyspaceByID(ctx context.Context, keyspaceID uint32
 	}
 
 	k.keyspaceMu.Lock()
-	defer k.keyspaceMu.Unlock()
 	meta := k.keyspaceIDMap[keyspaceID]
+	k.keyspaceMu.Unlock()
 	if meta != nil {
 		return meta, nil
 	}
@@ -117,6 +124,17 @@ func (k *keyspaceManager) GetKeyspaceByID(ctx context.Context, keyspaceID uint32
 		}
 		return nil
 	}, retry.WithBackoffBaseDelay(500), retry.WithBackoffMaxDelay(1000), retry.WithMaxTries(6))
+	if err != nil {
+		log.Error("retry to load keyspace from pd", zap.Uint32("keyspaceID", keyspaceID), zap.Error(err))
+		return nil, errors.Trace(err)
+	}
+
+	k.keyspaceMu.Lock()
+	defer k.keyspaceMu.Unlock()
+	// Double check, another goroutine might have fetched and stored it.
+	if meta, ok := k.keyspaceIDMap[keyspaceID]; ok {
+		return meta, nil
+	}
 
 	k.keyspaceMap[meta.Name] = meta
 	k.keyspaceIDMap[keyspaceID] = meta

--- a/pkg/keyspace/keyspace_manager.go
+++ b/pkg/keyspace/keyspace_manager.go
@@ -34,7 +34,7 @@ import (
 
 type KeyspaceManager interface {
 	LoadKeyspace(ctx context.Context, keyspace string) (*keyspacepb.KeyspaceMeta, error)
-	GetKeyspaceByID(keyspaceID uint32) *keyspacepb.KeyspaceMeta
+	GetKeyspaceByID(ctx context.Context, keyspaceID uint32) (*keyspacepb.KeyspaceMeta, error)
 	GetStorage(keyspace string) (kv.Storage, error)
 	Close()
 }
@@ -94,16 +94,34 @@ func (k *keyspaceManager) LoadKeyspace(ctx context.Context, keyspace string) (*k
 	return meta, nil
 }
 
-func (k *keyspaceManager) GetKeyspaceByID(keyspaceID uint32) *keyspacepb.KeyspaceMeta {
+func (k *keyspaceManager) GetKeyspaceByID(ctx context.Context, keyspaceID uint32) (*keyspacepb.KeyspaceMeta, error) {
 	if kerneltype.IsClassic() {
 		return &keyspacepb.KeyspaceMeta{
 			Name: common.DefaultKeyspace,
-		}
+		}, nil
 	}
 
 	k.keyspaceMu.Lock()
 	defer k.keyspaceMu.Unlock()
-	return k.keyspaceIDMap[keyspaceID]
+	meta := k.keyspaceIDMap[keyspaceID]
+	if meta != nil {
+		return meta, nil
+	}
+
+	var err error
+	pdAPIClient := appcontext.GetService[pdutil.PDAPIClient](appcontext.PDAPIClient)
+	err = retry.Do(ctx, func() error {
+		meta, err = pdAPIClient.GetKeyspaceMetaByID(ctx, keyspaceID)
+		if err != nil {
+			return err
+		}
+		return nil
+	}, retry.WithBackoffBaseDelay(500), retry.WithBackoffMaxDelay(1000), retry.WithMaxTries(6))
+
+	k.keyspaceMap[meta.Name] = meta
+	k.keyspaceIDMap[keyspaceID] = meta
+
+	return meta, nil
 }
 
 func (k *keyspaceManager) GetStorage(keyspace string) (kv.Storage, error) {

--- a/pkg/pdutil/api_client.go
+++ b/pkg/pdutil/api_client.go
@@ -37,6 +37,7 @@ import (
 	"github.com/pingcap/ticdc/pkg/retry"
 	"github.com/pingcap/ticdc/pkg/security"
 	pd "github.com/tikv/pd/client"
+	pdhttp "github.com/tikv/pd/client/http"
 	"go.uber.org/zap"
 )
 
@@ -104,13 +105,15 @@ type PDAPIClient interface {
 	Healthy(ctx context.Context, endpoint string) error
 	ScanRegions(ctx context.Context, span heartbeatpb.TableSpan) ([]RegionInfo, error)
 	LoadKeyspace(ctx context.Context, name string) (*keyspacepb.KeyspaceMeta, error)
+	GetKeyspaceMetaByID(ctx context.Context, keyspaceID uint32) (*keyspacepb.KeyspaceMeta, error)
 	Close()
 }
 
 // pdAPIClient is the api client of Placement Driver, include grpc client and http client.
 type pdAPIClient struct {
-	grpcClient pd.Client
-	httpClient *httputil.Client
+	grpcClient   pd.Client
+	httpClient   *httputil.Client
+	pdHttpClient pdhttp.Client
 }
 
 // NewPDAPIClient create a new pdAPIClient.
@@ -119,9 +122,24 @@ func NewPDAPIClient(pdClient pd.Client, conf *security.Credential) (PDAPIClient,
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+
+	discovery := pdClient.GetServiceDiscovery()
+	pdhttpOpts := make([]pdhttp.ClientOption, 0)
+
+	tlsConf, err := conf.ToTLSConfigWithVerify()
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsConf != nil {
+		opt := pdhttp.WithTLSConfig(tlsConf)
+		pdhttpOpts = append(pdhttpOpts, opt)
+	}
+
 	return &pdAPIClient{
-		grpcClient: pdClient,
-		httpClient: dialClient,
+		grpcClient:   pdClient,
+		httpClient:   dialClient,
+		pdHttpClient: pdhttp.NewClientWithServiceDiscovery("cdc", discovery, pdhttpOpts...),
 	}, nil
 }
 
@@ -281,6 +299,14 @@ func (pc *pdAPIClient) LoadKeyspace(ctx context.Context, name string) (*keyspace
 		return &keyspacepb.KeyspaceMeta{}, nil
 	}
 	return pc.grpcClient.LoadKeyspace(ctx, name)
+}
+
+func (pc *pdAPIClient) GetKeyspaceMetaByID(ctx context.Context, keyspaceID uint32) (*keyspacepb.KeyspaceMeta, error) {
+	if kerneltype.IsClassic() {
+		return &keyspacepb.KeyspaceMeta{}, nil
+	}
+
+	return pc.pdHttpClient.GetKeyspaceMetaByID(ctx, keyspaceID)
 }
 
 // ServiceSafePoint contains gc service safe point

--- a/pkg/pdutil/api_client.go
+++ b/pkg/pdutil/api_client.go
@@ -125,7 +125,7 @@ func NewPDAPIClient(pdClient pd.Client, conf *security.Credential) (PDAPIClient,
 
 	pdHttpClient, err := newPdHttpClient(pdClient, conf)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, err
 	}
 
 	return &pdAPIClient{
@@ -141,7 +141,7 @@ func newPdHttpClient(pdClient pd.Client, conf *security.Credential) (pdhttp.Clie
 
 	tlsConf, err := conf.ToTLSConfigWithVerify()
 	if err != nil {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
 
 	if tlsConf != nil {

--- a/pkg/pdutil/api_client.go
+++ b/pkg/pdutil/api_client.go
@@ -125,7 +125,7 @@ func NewPDAPIClient(pdClient pd.Client, conf *security.Credential) (PDAPIClient,
 
 	pdHttpClient, err := newPdHttpClient(pdClient, conf)
 	if err != nil {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
 
 	return &pdAPIClient{

--- a/pkg/pdutil/api_client.go
+++ b/pkg/pdutil/api_client.go
@@ -123,6 +123,19 @@ func NewPDAPIClient(pdClient pd.Client, conf *security.Credential) (PDAPIClient,
 		return nil, errors.Trace(err)
 	}
 
+	pdHttpClient, err := newPdHttpClient(pdClient, conf)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &pdAPIClient{
+		grpcClient:   pdClient,
+		httpClient:   dialClient,
+		pdHttpClient: pdHttpClient,
+	}, nil
+}
+
+func newPdHttpClient(pdClient pd.Client, conf *security.Credential) (pdhttp.Client, error) {
 	discovery := pdClient.GetServiceDiscovery()
 	pdhttpOpts := make([]pdhttp.ClientOption, 0)
 
@@ -136,11 +149,8 @@ func NewPDAPIClient(pdClient pd.Client, conf *security.Credential) (PDAPIClient,
 		pdhttpOpts = append(pdhttpOpts, opt)
 	}
 
-	return &pdAPIClient{
-		grpcClient:   pdClient,
-		httpClient:   dialClient,
-		pdHttpClient: pdhttp.NewClientWithServiceDiscovery("cdc", discovery, pdhttpOpts...),
-	}, nil
+	client := pdhttp.NewClientWithServiceDiscovery("cdc", discovery, pdhttpOpts...)
+	return client, nil
 }
 
 // Close the pd api client, at the moment only close idle http connections if there is any.

